### PR TITLE
feat(analytics): auto-provision GA4 custom dimensions

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Automatically provisions Newspack's standard set of GA4 event-scoped custom dimensions (27 total, covering Gate Intelligence, reader identity, prompt attribution, checkout, and content context) on the publisher's connected GA4 property.

Triggered automatically when Site Kit's `googlesitekit_analytics-4_settings` option gains a `propertyID` or the property changes, via a single-shot `wp_schedule_single_event` keyed on the property ID. Also exposed as `wp newspack ga4-dimensions provision[--dry-run]`. Provisioning is idempotent (existing parameter names are skipped) and per-dimension failures are logged but don't abort the run. The `newspack_ga4_dimensions_provisioned` option stores a summary of the last run keyed on property ID.

### Auth source: Newspack OAuth primary, Site Kit fallback

Site Kit's Analytics module only requests `analytics.readonly` by default; `analytics.edit` is granted on-demand for SK-initiated write actions, so publishers who connected an _existing_ GA4 property via Site Kit hit a 403 on the Admin API. Newspack's own Google OAuth (`_newspack_google_oauth`) already has `analytics.edit` in `Google_OAuth::REQUIRED_SCOPES`, so `GA4_Custom_Dimensions::with_admin_client()` now prefers it (via a raw-HTTP client against `analyticsadmin.googleapis.com/v1beta`) and falls back to Site Kit when Newspack OAuth isn't configured or the call errors. Active path is surfaced as an `Auth source:` line in CLI output and in the summary option. Prerequisite: `analyticsadmin.googleapis.com` enabled on the Newspack OAuth proxy's GCP project (done, prod + staging). Edge case: sites with neither Newspack OAuth nor Site Kit edit-scope still 403 – they need to reconnect Site Kit to grant edit.

Closes NPPM-2738.

### How to test the changes in this Pull Request:

1. On a site with Site Kit active but not yet connected to GA4, run `wp newspack ga4-dimensions provision --dry-run`. It should error out with "No GA4 property ID configured in Site Kit."
2. Connect Site Kit to a GA4 property. Confirm a `newspack_ga4_provision_dimensions` scheduled action appears (`wp cron event list | grep ga4`); wait for or manually trigger it. Verify all 27 Newspack dimensions exist in GA4 (Admin → Property → Custom definitions → Custom dimensions).
3. Re-run `wp newspack ga4-dimensions provision` and `... --dry-run`. Both should be no-ops (0 missing, every dimension reports as already existing).
4. **Auth-source selection.** On a site with Newspack OAuth connected (`wp option get _newspack_google_oauth` returns a populated option) and Site Kit in readonly mode, run `wp newspack ga4-dimensions provision --dry-run`. Output should include `Auth source: newspack` and successfully list existing event-scoped dimensions. On a site with Site Kit in edit mode and no Newspack OAuth, the same command should print `Auth source: sitekit`.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/plugins/google-site-kit/class-ga4-custom-dimensions.php [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+		if ( ! get_current_user_id() ) {
+			$settings = get_option( 'googlesitekit_analytics-4_settings', [] );
+			$owner_id = isset( $settings['ownerID'] ) ? (int) $settings['ownerID'] : 0;
+			if ( $owner_id > 0 ) {
+				wp_set_current_user( $owner_id );
+			} else {
+				return new \WP_Error( 'newspack_ga4_dimensions', 'No Site Kit module owner found to authenticate as.' );
+			}
+		}
+		return new GoogleSiteKitAnalytics( new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE ) );
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-ga4-custom-dimensions.php]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

File: includes/plugins/google-site-kit/class-ga4-custom-dimensions.php [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+		$summary = [
+			'property_id'    => $property_id,
+			'timestamp'      => time(),
+			'created'        => $created,
+			'skipped_exists' => $skipped_exists,
+			'errors'         => $errors,
+		];
+
+		$previous = get_option( self::PROVISIONED_OPTION, [] );
+		if ( is_array( $previous ) && isset( $previous['created'] ) && is_array( $previous['created'] ) ) {
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-ga4-custom-dimensions.php]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

File: includes/plugins/google-site-kit/class-ga4-custom-dimensions.php [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+	 * Fired when Site Kit's analytics-4 option is first added. Schedules
+	 * provisioning if the new settings include a property ID.
+	 *
+	 * @param string $option Option name.
+	 * @param mixed  $value  New option value.
+	 */
+	public static function on_sitekit_settings_added( $option, $value ) {
+		if ( ! empty( $value['propertyID'] ) ) {
+			self::schedule_provisioning( (string) $value['propertyID'] );
+		}
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-ga4-custom-dimensions.php]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

File: includes/plugins/google-site-kit/class-googlesitekitanalytics.php [hunk 4]
```diff
@@ -81,4 +82,53 @@ public function get_ga4_settings( $account_id ) {
...
+				->properties_customDimensions // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+				->listPropertiesCustomDimensions( 'properties/' . $property_id, $params );
+			foreach ( $response['customDimensions'] as $dimension ) {
+				$dimensions[] = [
+					'name'          => $dimension['name'],
+					'parameterName' => $dimension['parameterName'],
+					'displayName'   => $dimension['displayName'],
+					'scope'         => $dimension['scope'],
+				];
+			}
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-googlesitekitanalytics.php:112]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

File: includes/plugins/google-site-kit/class-ga4-custom-dimensions.php [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+			'provisioned_option'    => get_option( self::PROVISIONED_OPTION, null ),
+		];
+	}
+
+	/**
+	 * Provision Newspack's standard GA4 custom dimensions.
+	 *
+	 * Idempotent: existing dimensions are detected and skipped. Creates
+	 * missing dimensions in priority order, stopping if the property's 50
+	 * event-scoped dimension limit would be exceeded.
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-ga4-custom-dimensions.php]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

File: includes/plugins/google-site-kit/class-ga4-custom-dimensions.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/plugins/google-site-kit/class-ga4-custom-dimensions.php]: Addressed in fe7593d92434b248036a70aa663382e4c477ea13.

**Comment by leogermani** [comment]
The dry run works, but when trying to actually create the dimensions with my standard sitekit connection, I get a bunch of 

```
  "error": {
    "code": 403,
    "message": "Request had insufficient authentication scopes.",
    "errors": [
      {
        "message": "Insufficient Permission",
        "domain": "global",
        "reason": "insufficientPermissions"
      }
    ],
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "ACCESS_TOKEN_SCOPE_INSUFFICIENT",
        "domain": "googleapis.com",
        "metadata": {
          "service": "analyticsadmin.googleapis.com",
          "method": "google.analytics.admin.v1beta.AnalyticsAdminService.CreateCustomDimension"
        }
      }
    ]
  }
```

**Comment by adekbadek** [comment]
Context on the 403 `ACCESS_TOKEN_SCOPE_INSUFFICIENT`:

Site Kit's Analytics module requests only `READONLY_SCOPE` (`Analytics_4::get_scopes()`); `EDIT_SCOPE` is granted incrementally, only on SK write actions (`create-property`, `create-custom-dimension`, `enhanced-measurement-settings`, etc.). Publishers who connected an _existing_ GA4 property never triggered that prompt.

Newspack's own OAuth already holds `analytics.edit` via `Google_OAuth::REQUIRED_SCOPES`, so preferring it avoids asking publishers to reconnect SK. Newspack previously provisioned custom dimensions through this OAuth, but against Universal Analytics' Management API (retired with UA). GA4's Admin API is a separate GCP product – same OAuth scope, different enablement toggle. `analyticsadmin.googleapis.com` is now enabled on the Newspack OAuth proxy's GCP project (prod + staging); no user reauth required.

Edge case: sites with neither Newspack OAuth nor SK edit still 403 – they need to reconnect Site Kit.

@leogermani – you can run the following snippet to check if you have either auth:

```
wp option get _newspack_google_oauth >/dev/null 2>&1 && np=YES || np=NO; wp db query "SELECT 1 FROM wp_usermeta WHERE meta_key='googlesitekit_auth_granted_scopes' AND meta_value LIKE '%analytics.edit%' LIMIT 1" --skip-column-names 2>/dev/null | grep -q 1 && sk=YES || sk=NO; echo "SK:$sk | NP:$np"
```

**Review by dkoo — CHANGES_REQUESTED** [review]
Manual testing worked for me as described—the cron job gets scheduled when I connect Analytics, and successfully creates the missing custom dimensions when it runs. CLI also successfully checks for the dimensions and reports 0 missing after this.

One thing that's missing here is some kind of periodic recheck of provisioned dimensions. For sites that are already connected to Site Kit/Analytics, the cron job is scheduled once to check provisioned dimensions. After the provisioning completes, the local `newspack_ga4_provision_dimensions` option is updated with the custom dimensions that are provisioned on the connected Analytics property at that moment, but this option is never updated unless the provisioner is manually run again. If we were to change anything in our list of custom dimensions, or someone deletes custom dimensions in Analytics, the local option won't be aware of these changes and the custom dimensions won't be reprovisioned unless we manually run the CLI command. Should we consider scheduling an infrequent recheck (maybe once per month) to ensure that the provisioned dimensions remain healthy over time? If the dimensions are healthy then the check will be no-op and cheap to run so infrequently.

Other concerns surfaced by Claude (including in this top-level comment due to platform-wide problems with inline PR comments on GitHub right now):

### 🔴 No tests on a path that mutates third-party state

The provisioner:

- mutates an external Google project (creates dimensions that count against GA4's 50-event-scoped cap),
- juggles two distinct auth sources with different failure modes,
- runs from cron with switched user context.

Easily-testable shapes that would catch real regressions:

- **`with_admin_client` fallback selection.** Mock `Google_OAuth_GA4_Client::build()` returning a stub; verify (a) `$source === 'newspack'` when Newspack OAuth is configured, (b) source flips to `'sitekit'` when Newspack callback throws, (c) source flips when callback returns `WP_Error`, (d) early-returns `WP_Error` when neither is available.
- **`schedule_provisioning` dedup.** Same property → no schedule. Different property → schedule. Already-scheduled → no schedule. Already-provisioned for property X but property is now Y → schedule.
- **Property-switch summary merge.** Pre-seed `PROVISIONED_OPTION` with `{ property_id: 'A', created: [...] }`; run `provision()` against property B; assert previous `created` is *not* merged into the new summary.
- **Idempotent provision.** Stub list to return all 27 dimensions; assert zero `create_custom_dimension` calls and `created: []`.

These are mock-shaped tests, not integration tests against GA4. ~150 lines and they protect the sharpest edges.

### 🟡 Newspack OAuth path doesn't verify scope before calling

`Google_OAuth_GA4_Client::build()` returns a client whenever `is_oauth_configured() && get_oauth2_credentials()` succeeds — it doesn't check that the stored token actually carries `analytics.edit`. If a publisher's stored token was issued before `analytics.edit` was added to `Google_OAuth::REQUIRED_SCOPES` (or a publisher revoked the scope from their Google account), the Newspack path will return a 401/403 from the Admin API.

The fallback chain *does* handle this — the exception is caught and Site Kit is tried — but:

1. The error is buried in `Logger::log` and not surfaced to the CLI summary as "Newspack path failed because token lacks scope X."
2. Publishers in this state silently fall back to a Site Kit path that probably also lacks `analytics.edit`, ending up in @leogermani's original 403 with no clear diagnosis.

Cheap improvement: in `with_admin_client`, before invoking the Newspack callback, check `$auth_data['scope']` (or `$auth_data['granted_scopes']`, whichever the proxy persists) for `analytics.edit`. If absent, skip Newspack and try Site Kit directly. The CLI output's `Auth source: newspack (token lacks edit scope)` will surface the missing required scope as the cause of 401/403 responses.

### 🟡 Cron-execution time budget vs. dimension count

`wp_remote_request` has a 15s timeout (good). 27 dimensions × up to 15s = up to 6.75 min worst case. WP-Cron runs synchronously inside the request that triggers it. PHP's `max_execution_time` in cron context is typically 30–60s. On a slow Google response, the cron request can be killed before `update_option( PROVISIONED_OPTION, $summary )` writes, leaving the site half-provisioned with no summary.

Real-world latency is much better than 15s/dimension, so this is unlikely in practice. But:
- `set_time_limit( 0 )` at the top of `provision()` (CLI gets this for free; cron does not), or
- Persist progress incrementally (write `$summary` after each successful create) so a partial run resumes cleanly,

would close the gap. The comment from @leogermani's 403 actually exercised this path — many failed creates each waiting on Google's full timeout. With 50 dimensions in flight, that hit was real.

### 🟡 `wp_next_scheduled` ignores args (non-blocking)

`wp_schedule_single_event` is called without `$args`, and `wp_next_scheduled( self::PROVISION_ACTION )` is checked without `$args`. So if the property changes A→B before A's event fires, the B change is dropped. The handler `provision()` reads `get_property_id()` at run time, which means whichever property is current at the firing moment wins — usually B, since the older A event runs the new B value. Net outcome: last-write-wins, which is what you want. But it means no record of the intermediate A provisioning attempt and any *partially* provisioned dimensions on A get orphaned with no summary entry (because the merge only carries forward when `property_id` matches).

Acceptable as designed. Worth a docblock note.

### 🟢 Minor (non-blocking)

- **`GoogleSiteKitAnalytics::list_custom_dimensions`** reads `$dimension['name']` etc. without `isset` guards, unlike the Newspack-path equivalent which does. Should match Copilot's earlier fix style for consistency.
- **`update_option( self::PROVISIONED_OPTION, $summary )`** — autoload defaults to `yes`. Summary is small (~27 entries). Fine, but `update_option(…, $summary, false )` would be the cleaner default for an option only read by CLI / status calls.
- **Two near-identical client implementations** (raw HTTP vs. Site Kit's wrapper). If the dimension list grows, or you add `update_custom_dimension` later, both must be touched. A thin interface (`GA4DimensionsClient`) implemented twice would document the contract. Not blocking.
- **CLI run without `--user=admin`** will silently fail the Newspack path because `Google_OAuth::permissions_check()` requires `current_user_can('manage_options')`. The PR's `with_admin_client` only switches user when `get_current_user_id() === 0`, but in WP-CLI the default is 0, so the `ownerID` switch kicks in. ✓ Verified: `wp_set_current_user( $owner_id )` runs and the path works. Worth a sentence in the test plan or `--help` output.
- **CLI `provision` output** doesn't show errors prominently — they're after the `WP_CLI::warning( 'Errors:' )` log call but before `WP_CLI::success(...)`. A non-zero exit code on partial failure would help CI/Ansible-driven runs detect the problem. Currently a half-failed run still exits 0.
- **27 × 1 + Newspack `analytics.edit`** is fine, but if `get_dimensions()` ever crosses the 50 event-scoped cap, the priority ordering claim in the docblock becomes load-bearing. A unit test that pins `count(get_dimensions()) <= 50` is one line and worth having.

**Comment by adekbadek** [comment]
Addressed in 8bf25b09e.

**Periodic recheck** — added a monthly recheck (Action Scheduler, group `newspack`) that re-runs `provision()`; it self-heals dimension-list additions and GA4-side deletions and is a no-op (one list call, zero writes) when everything's in place. Scheduled on connect / property change, on `admin_init` for sites that were already connected before this shipped, and after a CLI run; unscheduled when the property is disconnected.

**Tests** — new `tests/unit-tests/ga4-custom-dimensions.php` (`@group ga4-dimensions`, HTTP-mocked, no real Google project): auth-route selection (Newspack vs Site Kit fallback on error, skip-when-no-edit-scope, WP_Error when neither available), `schedule_provisioning` dedup (4 cases + non-array tolerance), idempotent provision (zero creates, no POSTs), property-switch vs same-property summary merge, GA4 client list/pagination/create/error, and the `count(get_dimensions()) <= 50` cap.

**Scope pre-check + diagnostics** — `Google_OAuth_GA4_Client::has_edit_scope()` (via a new `Google_OAuth::token_has_scope()` tokeninfo lookup); `with_admin_client` skips the Newspack route up front when the stored token lacks `analytics.edit` (so it doesn't fire 27 doomed creates), and on total failure the WP_Error / CLI output now names each route tried and why — so the original 403 reads as "token lacks analytics.edit / Site Kit not available" instead of a bare API error.

**Cron time budget** — `set_time_limit( 0 )` at the top of `provision()` (cron/AS run synchronously inside a 30–60s request); docblock notes that on hosts that disable it a cut-short run is safe because the next run lists what exists and only creates the remainder.

**`wp_next_scheduled` args** — docblock note on the last-write-wins behavior.

**Nits** — `GoogleSiteKitAnalytics::list_custom_dimensions` now `isset`-guards its reads (matches the Newspack-path client); `update_option( …, $summary, false )`; the CLI command exits non-zero when any dimension fails; `--user` is documented in `wp newspack ga4-dimensions provision --help`.

**Review by dkoo — APPROVED** [review]
